app/com/arpnetworking/metrics/portal/scheduling/JobExecutionRepository.java

```diff
@@ -0,0 +1,93 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling;
+
+import models.internal.Organization;
+import models.internal.scheduling.Job;
+import models.internal.scheduling.JobExecution;
+
+import java.time.Instant;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A storage medium for {@link JobExecution}s.
+ *
+ * @param <T> The type of result produced by the {@link Job}s.
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public interface JobExecutionRepository<T> {
+
+    /**
+     * Open / connect to the repository. Must be called before any other methods.
+     */
+    void open();
+
+    /**
+     * Close the repository. Any further operations are illegal until the next open() call.
+     */
+    void close();
+
+    /**
+     * Get the last successful execution, if any.
+     *
+     * @param jobId The UUID of the job that completed.
+     * @param organization The organization owning the job.
+     * @throws NoSuchElementException if no job has the given UUID.
+     * @return The last successful execution.
+     */
+    Optional<JobExecution.Success<T>> getLastSuccess(UUID jobId, Organization organization) throws NoSuchElementException;
+
+    /**
+     * Get the last completed execution, regardless of if it succeeded.
+     *
+     * @param jobId The UUID of the job that completed.
+     * @param organization The organization owning the job.
+     * @throws NoSuchElementException if no job has the given UUID.
+     * @return The last completed execution.
+     */
+    Optional<JobExecution<T>> getLastCompleted(UUID jobId, Organization organization) throws NoSuchElementException;
+
+    /**
+     * Notify the repository that a job has started executing.
+     *
+     * @param jobId The UUID of the job that completed.
+     * @param organization The organization owning the job.
+     * @param scheduled The time that the job started running for.
+     */
+    void jobStarted(UUID jobId, Organization organization, Instant scheduled);
+
+    /**
+     * Notify the repository that a job finished executing successfully.
+     *
+     * @param jobId The UUID of the job that completed.
+     * @param organization The organization owning the job.
+     * @param scheduled The time that the completed job-run was scheduled for.
+     * @param result The result that the job computed.
+     */
+    void jobSucceeded(UUID jobId, Organization organization, Instant scheduled, T result);
```

Comment:
If you wanted to get _crazy_, you could make this take a `JobExecution.Started<T>` as proof that the job has already started.

Recent unpleasant experiences have mellowed me on the whole "use the typechecker to verify your program's correctness" thing, though, and I'll happily admit that that's _probably_ going overboard.

Reply:
I agree - while that would be nice it would put more of a burden on the user of the API than I'd like, especially since this is all on top of the database so we'd essentially be _requiring_ an initial lookup if you started with just a Job ID.

It seems easier / more idiomatic to handle that precondition within the repository implementation and throw if its violated.